### ISSUE TITLE
Python test failures with pybind11 v2.8.1

### ISSUE THREAD

**Issue by Infinoid** [halide/Halide#6524]
Debian bookworm, by default, has llvm 13, python 3.9.9, and pybind11 2.8.1.  Two python tests fail:

```
30/42 Test #579: python_tutorial_lesson_04_debugging_2 ......................Subprocess aborted***Exception:   0.33 sec
terminate called after throwing an instance of 'std::runtime_error'
  what():  pybind11::object_api<>::operator() PyGILState_Check() failure.

      Start 580: python_tutorial_lesson_05_scheduling_1
31/42 Test #580: python_tutorial_lesson_05_scheduling_1 .....................Subprocess aborted***Exception:   0.55 sec
<...>
Evaluating gradient over a 5x4 box with x split by two 
Begin pipeline gradient_split_5x4.0()
produce gradient_split_5x4:
  for y:
    for x.x_outer:
      for x.x_inner in [0, 1]:
        gradient_split_5x4(...) = ...
terminate called after throwing an instance of 'std::runtime_error'
  what():  pybind11::object_api<>::operator() PyGILState_Check() failure.
```

When I revert pybind11 to version 2.6.2 (using packages from the previous debian release), reconfigure and rebuild, the python tests pass.

```
1/2 Test #571: python_tutorial_lesson_04_debugging_2 ....   Passed    0.45 sec
2/2 Test #572: python_tutorial_lesson_05_scheduling_1 ...   Passed    6.89 sec
```

Here is a minimal reproducer, based on tutorial lesson 04:

```python3
#!/usr/bin/env python3

import halide as hl

x = hl.Var("x")

test = hl.Func("test")
test[x] = 0
test.trace_stores()
test.parallel(x)
test.realize([8])
```

If I comment out the call to `trace_stores()` or the call to `parallel()`, the test passes.  Both are necessary to trigger the bug.

**Comment by Infinoid** [comment]
pybind/pybind11#2919 was added in pybind11 v2.7.0, I think Halide runs afoul of that.  Maybe it's calling back into python for tracing?

**Comment by Infinoid** [comment]
Note, it fails with cmake `-DCMAKE_BUILD_TYPE=Debug`, but passes with `-DCMAKE_BUILD_TYPE=Release`.  Release defines `NDEBUG`, which disables pybind11's sanity check.

**Comment by Infinoid** [comment]
Here's a backtrace from `lesson_04_debugging_2.py`:
```
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:49
#1  0x00007ffff7c5c536 in __GI_abort () at abort.c:79
#2  0x00007ffff062687a in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ffff0631ffa in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff0632065 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff06322f9 in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007ffff7484757 in pybind11::pybind11_fail (
    reason=0x7ffff783a670 "pybind11::object_api<>::operator() PyGILState_Check() failure.") at /usr/include/pybind11/detail/common.h:838
#7  0x00007ffff755d7f6 in pybind11::detail::object_api<pybind11::detail::accessor<pybind11::detail::accessor_policies::str_attr> >::operator()<(pybind11::return_value_policy)1, pybind11::tuple&> (this=0x7fffef4e5980) at /usr/include/pybind11/cast.h:1397
#8  0x00007ffff755cfc5 in pybind11::detail::print (args=..., kwargs=...) at /usr/include/pybind11/pybind11.h:2273
#9  0x00007ffff755dca4 in pybind11::print<(pybind11::return_value_policy)1, char const*&, pybind11::arg_v> ()
    at /usr/include/pybind11/pybind11.h:2302
#10 0x00007ffff755be15 in Halide::PythonBindings::(anonymous namespace)::halide_python_print (
    msg=0x7fffef4e5bc8 "Store test.0(4) = 0\n") at /src/python_bindings/src/PyError.cpp:13
#11 0x00007ffff153a221 in Halide::Internal::(anonymous namespace)::print_handler (context=0x7fffffffd4e8, 
    msg=0x7fffef4e5bc8 "Store test.0(4) = 0\n") at /src/src/JITModule.cpp:514
#12 0x00007ffff7fc0ace in halide_default_trace ()
#13 0x00007ffff153a444 in Halide::Internal::(anonymous namespace)::trace_handler (context=0x7fffffffd4e8, e=0x7fffef4e6ca8)
    at /src/src/JITModule.cpp:570
#14 0x00007ffff7fc1e8b in halide_trace_helper ()
#15 0x00007ffff04eb056 in test.par_for.test.s0.x.rebased ()
#16 0x0000000000000001 in ?? ()
#17 0x0000000000000001 in ?? ()
#18 0x0000000000000000 in ?? ()
```

**Comment by Infinoid** [comment]
So:

* Python calls a Halide-JIT-generated function , which runs with the GIL held. 
* Halide runtime spawns worker threads.
* The worker threads try to call pybind11's py::print function to emit traces.
* Pybind11 complains, correctly, that the worker thread doesn't hold the GIL.

Trying to acquire the GIL hangs, because the main thread is still holding it.
I tried teaching the main thread to release the GIL (as suggested in #5631), but I still saw hangs when I tried this.

Here are two workarounds:
1. downgrade pybind11 to <2.7.0
1. comment out [the assignment of `handlers.custom_print` in PyError.cpp](https://github.com/halide/Halide/blob/9a530b1960c2f9832cff7bc54b54d9a0ea12c398/python_bindings/src/PyError.cpp#L36)
